oxide-auth-iron/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+        serde_urlencoded::from_str(self.0.url.query().unwrap_or(""))
+            .map_err(|_| Error::BadRequest)
+            .map(Cow::Owned)
+    }
+
+    fn urlbody(&mut self) -> Result<Cow<dyn QueryParameter + 'static>, Self::Error> {
+        let content_type = self.0.headers.get::<headers::ContentType>();
+        let formatted = content_type
+            .map(|ct| ct == &headers::ContentType::form_url_encoded())
+            .unwrap_or(false);
+        if !formatted {
+            return Err(Error::BadRequest);
+        }
+
+        serde_urlencoded::from_reader(&mut self.0.body)
+            .map_err(|_| Error::BadRequest)
+            .map(Cow::Owned)
+    }
+
+    fn authheader(&mut self) -> Result<Option<Cow<str>>, Self::Error> {
+        // Get the raw header.
+        match self.0.headers.get::<headers::Authorization<String>>() {
+            None => Ok(None),
+            Some(header) => Ok(Some(Cow::Borrowed(&header.0))),
+        }
+    }
+}
+
+impl WebResponse for OAuthResponse {
+    type Error = Error;
+
+    fn ok(&mut self) -> Result<(), Self::Error> {
+        self.0.status = Some(Status::Ok);
+        Ok(())
+    }
+
+    fn redirect(&mut self, url: Url) -> Result<(), Self::Error> {
+        self.0.status = Some(Status::Found);
+        self.0.headers.set(headers::Location(url.into_string()));
+        Ok(())
+    }
+
+    fn client_error(&mut self) -> Result<(), Self::Error> {
+        self.0.status = Some(Status::BadRequest);
+        Ok(())
+    }
+
+    fn unauthorized(&mut self, header_value: &str) -> Result<(), Self::Error> {
+        self.0.status = Some(Status::Unauthorized);
+        let value_owned = header_value.as_bytes().to_vec();
+        self.0
+            .headers
+            .set_raw("WWW-Authenticate", vec![value_owned]);
+        Ok(())
+    }
+
+    fn body_text(&mut self, text: &str) -> Result<(), Self::Error> {
+        self.0.headers.set(headers::ContentType::plaintext());
+        self.0.body = Some(Box::new(text.to_string()));
+        Ok(())
+    }
+
+    fn body_json(&mut self, data: &str) -> Result<(), Self::Error> {
+        self.0.headers.set(headers::ContentType::json());
+        self.0.body = Some(Box::new(data.to_string()));
+        Ok(())
+    }
+}
+
+impl<'a, 'b, 'c: 'b> From<SimpleError<OAuthRequest<'a, 'b, 'c>>> for OAuthError {
+    fn from(error: SimpleError<OAuthRequest<'a, 'b, 'c>>) -> Self {
+        let as_oauth = match error {
+            SimpleError::Web(Error::BadRequest) => EndpointError::BadRequest,
+            SimpleError::OAuth(oauth) => oauth,
+        };
+
+        let status = match as_oauth {
+            EndpointError::BadRequest => Status::BadRequest,
+            EndpointError::DenySilently => Status::BadRequest,
+            EndpointError::PrimitiveError => Status::InternalServerError,
+        };
+
+        OAuthError(IronError::new(as_oauth, status))
+    }
+}
+
+#[derive(Debug)]
+pub struct OAuthResponse(pub Response);
```

Comment:
Please reorder like the other. Also, this is missing a doc-comment.

 It would be great if there was `#![warn(missing_docs)]` in all individual libraries.